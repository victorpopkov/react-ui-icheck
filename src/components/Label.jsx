import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Label extends Component {
  constructor(props) {
    super(props);

    this.handleHover = this.handleHover.bind(this);
  }

  handleHover(event, isHovered = true) {
    const { disabled, handleHover, isMobile, labelHover } = this.props;

    if (disabled) return;

    if (handleHover && labelHover && !isHovered) {
      handleHover(event, false);
    } else if (handleHover) {
      handleHover(event, isHovered);
    }

    if (isMobile) event.stopPropagation();
  }

  renderLabel() {
    const { label, labelChildren, labelTag, labelTagClassName } = this.props;

    if (!React.isValidElement(label)) {
      const Tag = labelTag;

      if (labelChildren) {
        return <Tag className={labelTagClassName}>{labelChildren}</Tag>;
      }

      return (
        // eslint-disable-next-line react/no-danger
        <Tag
          className={labelTagClassName}
          dangerouslySetInnerHTML={{ __html: label }}
        />
      );
    }

    return null;
  }

  render() {
    const { children, labelClassName } = this.props;

    return (
      <label
        className={labelClassName || null}
        onBlur={(event) => this.handleHover(event, false)}
        onFocus={this.handleHover}
        onMouseOut={(event) => this.handleHover(event, false)}
        onMouseOver={this.handleHover}
        onTouchEnd={(event) => this.handleHover(event, false)}
        onTouchStart={this.handleHover}
      >
        {children}
        {this.renderLabel()}
      </label>
    );
  }
}

Label.propTypes = {
  children: PropTypes.node.isRequired,
  handleHover: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  isMobile: PropTypes.bool,
  label: PropTypes.node,
  labelChildren: PropTypes.node,
  labelClassName: PropTypes.string,
  labelHover: PropTypes.bool,
  labelTag: PropTypes.string,
  labelTagClassName: PropTypes.string,
};

Label.defaultProps = {
  disabled: false,
  isMobile: false,
  label: null,
  labelChildren: null,
  labelClassName: null,
  labelHover: true,
  labelTag: 'span',
  labelTagClassName: null,
};

export default Label;
