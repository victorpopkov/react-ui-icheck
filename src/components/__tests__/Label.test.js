import { mount, render, shallow } from 'enzyme';
import React from 'react';
import expect from 'expect';
import Label from '../Label'; // eslint-disable-line sort-imports

describe('<Label />', () => {
  let testLabel = null;

  beforeEach(() => {
    testLabel = <Label handleHover={() => false}>test</Label>;
  });

  it('should have the default props when mounted', () => {
    const wrapper = mount(testLabel);
    expect(wrapper.props()).toMatchSnapshot();
  });

  describe('should render', () => {
    it('the default React components', () => {
      const wrapper = shallow(testLabel);
      expect(wrapper).toMatchSnapshot();
    });

    it('the default static HTML', () => {
      const wrapper = render(testLabel);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
