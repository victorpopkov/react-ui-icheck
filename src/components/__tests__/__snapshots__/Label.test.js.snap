// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label /> should have the default props when mounted 1`] = `
Object {
  "children": "test",
  "disabled": false,
  "handleHover": [Function],
  "isMobile": false,
  "label": null,
  "labelChildren": null,
  "labelClassName": null,
  "labelHover": true,
  "labelTag": "span",
  "labelTagClassName": null,
}
`;

exports[`<Label /> should render the default React components 1`] = `
<label
  className={null}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
>
  test
  <span
    className={null}
    dangerouslySetInnerHTML={
      Object {
        "__html": null,
      }
    }
  />
</label>
`;

exports[`<Label /> should render the default static HTML 1`] = `
<label>
  test
  <span />
</label>
`;
